### PR TITLE
BZ1876855 doc storage exceptions to etcd restore

### PR DESCRIPTION
**Pull request by lpettyjo**
4.7+

https://bugzilla.redhat.com/show_bug.cgi?id=1876855

Adds information about storage exceptions for an etcd restore.

Preview: https://deploy-preview-33930--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/cluster-tasks.html#dr-scenario-cluster-state-issues_post-install-cluster-tasks

PTAL: @gnufied @duanwei33

### CODE REVIEW

File: modules/dr-restoring-cluster-state.adoc [hunk 1]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
 ----
 
 Note that it might take several minutes after completing this procedure for all services to be restored. For example, authentication by using `oc login` might not immediately work until the OAuth server pods are restarted.
+
+[id="dr-scenario-cluster-state-issues_{context}"]
+= Issues and workarounds for restoring persistent storage state
+
+If your {product-title} uses persistent storage of any form, typically a state of the world is stored outside etcd. It could be an ElasticSearch cluster running in a pod or a database running in a statefulset. When restoring from an etcd backup, the status of the workloads in {product-title} is also restored. However, if the etcd snapshot is old, the status might be invalid or outdated.
```
Comment (bobfuru) [inline modules/dr-restoring-cluster-state.adoc:321]:
```suggestion
If your {product-title} cluster uses persistent storage of any form, a state of the world is typically stored outside etcd. It might be an Elasticsearch cluster running in a pod or a database running in a `StatefulSet` object. When restoring from an etcd backup, the status of the workloads in {product-title} is also restored. However, if the etcd snapshot is old, the status might be invalid or outdated.
```
^Note that you'll need to manually apply these changes and not click the "Accept change" button.

Also, not sure how familiar users are with the term "state of the world"?

File: modules/dr-restoring-cluster-state.adoc [hunk 2]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
 ----
 
 Note that it might take several minutes after completing this procedure for all services to be restored. For example, authentication by using `oc login` might not immediately work until the OAuth server pods are restarted.
+
+[id="dr-scenario-cluster-state-issues_{context}"]
+= Issues and workarounds for restoring persistent storage state
```
Comment (bobfuru) [inline modules/dr-restoring-cluster-state.adoc:319]: "_a_ persistent storage state"?

File: modules/dr-restoring-cluster-state.adoc [hunk 3]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+If your {product-title} uses persistent storage of any form, typically a state of the world is stored outside etcd. It could be an ElasticSearch cluster running in a pod or a database running in a statefulset. When restoring from an etcd backup, the status of the workloads in {product-title} is also restored. However, if the etcd snapshot is old, the status might be invalid or outdated.
+
+[IMPORTANT]
+====
+The contents of PVs are never part of the etcd snapshot. When restoring an {product-title} cluster from an etcd snapshot, non-critical workloads might gain access to critical data (or vice-versa).
```
Comment (bobfuru) [inline modules/dr-restoring-cluster-state.adoc:325]: - You should spell out PV on first instance here but not later in the same module.

- s/data (or vice-versa)./data, or vice versa.

File: modules/dr-restoring-cluster-state.adoc [hunk 4]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+* Cloudprovider or storage provider credentials were updated after the etcd snapshot was taken. This causes any CSI drivers or operators that depend on the those credentials to not work, You might have to manually update the credentials required by those drivers or operators.
+
+* A device is removed or renamed from {product-title} nodes after the etcd snapshot is taken. Local Storage Operator creates symlinks for each persistent volume that it manages from `/dev/disk/by-id` or `/dev directories`. This situation might cause the local PVs to refer to devices that no longer exist.
++
+To fix this problem, an administror must:
+
+. Manually remove the PVs with invalid devices.
+. Remove symlink from respective nodes.
+. Delete LocalVolume or LocalVolumeSet objects (see _Storage_ -> _Configuring persistent storage_ -> _Persistent storage using local volumes_ -> _Deleting the Local Storage Operator Resources_).
```
Comment (bobfuru) [inline modules/dr-restoring-cluster-state.adoc:342]:
```suggestion
The following are some examples scenarios that produce an out-of-date status:

* A MySQL database is running in a pod backed up by a PV object. Restoring {product-title} from an etcd snapshot does not bring back the volume on the storage provider and does not produce a running MySQL pod, despite the pod repeatedly attempting to start. You must manually restore this pod by restoring the volume on the storage provider and then editing the PV to point to the new volume.

* Pod P1 is running using volume A, which is attached to node X. If the etcd snapshot is taken while another pod is using the same volume on node Y, then when the etcd restore is performed, pod P1 might not be able to start correctly due to the volume still being attached to node Y. {product-title} is aware of the attachment, and does not automatically detach it. When this occurs, the volume must be manually detached from node Y so that the volume can attach on node X, and then pod P1 can start.

* Cloudprovider or storage provider credentials were updated after the etcd snapshot was taken. This causes any CSI drivers or Operators that depend on the those credentials to not work. You might have to manually update the credentials required by those drivers or Operators.

* A device is removed or renamed from {product-title} nodes after the etcd snapshot is taken. The Local Storage Operator creates symlinks for each PV that it manages from `/dev/disk/by-id` or `/dev directories`. This situation might cause the local PVs to refer to devices that no longer exist.
+
To fix this problem, an administrator must:

. Manually remove the PVs with invalid devices.
. Remove symlinks from respective nodes.
. Delete `LocalVolume` or `LocalVolumeSet` objects (see _Storage_ -> _Configuring persistent storage_ -> _Persistent storage using local volumes_ -> _Deleting the Local Storage Operator Resources_).
```

**Review by bobfuru — COMMENTED** [review]
Some suggested edits. Note that the items in the "Suggested change" boxes will need to be applied manually. Thanks!

File: modules/dr-restoring-cluster-state.adoc (hunk 1 above)
Comment (lpettyjo) [reply modules/dr-restoring-cluster-state.adoc:321]: @gnufied consider Bob's comment about the "state of the world" term.

Comment (gnufied) [reply modules/dr-restoring-cluster-state.adoc:321]: we can reword it to say "state of the cluster" ?

File: modules/dr-restoring-cluster-state.adoc [hunk 5]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+====
+
+The following are some examples scenarios that produce an out-of-date status:
+
+* Mysql database is running in a pod backed up by a persistent volume (PV) object. Restoring {product-title} from an etcd snapshot does not bring back the volume on the storage provider, and does not produce a running Mysql pod, despite the pod repeatedly attempting to start. You must manually restore this pod by restoring the volume on the storage provider, and then editing the PV to point to the new volume.
+
+* Pod P1 is running using volume A, which is attached to node X. If the etcd snapshot is taken while another pod is using the same volume on node Y, then when the etcd restore is performed, pod P1 might not be able to start correctly due to the volume still being attached to node Y. {product-title} is aware of the attachment, and does not automatically detach it. When this occurs, the volume must be manually detached from node Y so that the volume can attach on node X, and then pod P1 can start.
```
Comment (gnufied) [inline modules/dr-restoring-cluster-state.adoc:332]: {product-title} is *not* aware

File: modules/dr-restoring-cluster-state.adoc [hunk 6]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+= Issues and workarounds for restoring a persistent storage state
+
+If your {product-title} cluster uses persistent storage of any form, a state of the world is typically stored outside etcd. It might be an Elasticsearch cluster running in a pod or a database running in a `StatefulSet`. When restoring from an etcd backup, the status of the workloads in {product-title} is also restored. However, if the etcd snapshot is old, the status might be invalid or outdated.
+
+[IMPORTANT]
+====
+The contents of persistent volumes (PVs) are never part of the etcd snapshot. When restoring an {product-title} cluster from an etcd snapshot, non-critical workloads might gain access to critical data, or vice-versa.
+====
+
+The following are some examples scenarios that produce an out-of-date status:
```
Comment (bobfuru) [inline modules/dr-restoring-cluster-state.adoc:328]: s/example/examples

File: modules/dr-restoring-cluster-state.adoc [hunk 7]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
 ----
 
 Note that it might take several minutes after completing this procedure for all services to be restored. For example, authentication by using `oc login` might not immediately work until the OAuth server pods are restarted.
+
+[id="dr-scenario-cluster-state-issues_{context}"]
+= Issues and workarounds for restoring a persistent storage state
+
+If your {product-title} cluster uses persistent storage of any form, a state of the world is typically stored outside etcd. It might be an Elasticsearch cluster running in a pod or a database running in a `StatefulSet`. When restoring from an etcd backup, the status of the workloads in {product-title} is also restored. However, if the etcd snapshot is old, the status might be invalid or outdated.
```
Comment (bobfuru) [inline modules/dr-restoring-cluster-state.adoc:321]: s/`StatefulSet`./`StatefulSet` object.
For actual API objects, we should follow the name with "object" per this OCP doc guideline: https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#object-references

File: modules/dr-restoring-cluster-state.adoc [hunk 8]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+* MySQL database is running in a pod backed up by a PV object. Restoring {product-title} from an etcd snapshot does not bring back the volume on the storage provider, and does not produce a running MySQL pod, despite the pod repeatedly attempting to start. You must manually restore this pod by restoring the volume on the storage provider, and then editing the PV to point to the new volume.
+
+* Pod P1 is running using volume A, which is attached to node X. If the etcd snapshot is taken while another pod is using the same volume on node Y, then when the etcd restore is performed, pod P1 might not be able to start correctly due to the volume still being attached to node Y. {product-title} is aware of the attachment, and does not automatically detach it. When this occurs, the volume must be manually detached from node Y so that the volume can attach on node X, and then pod P1 can start.
+
+* Cloudprovider or storage provider credentials were updated after the etcd snapshot was taken. This causes any CSI drivers or Operators that depend on the those credentials to not work, You might have to manually update the credentials required by those drivers or Operators.
```
Comment (bobfuru) [inline modules/dr-restoring-cluster-state.adoc:334]: - "Cloud provider" should be two words.
- There should be a period after "to not work"

File: modules/dr-restoring-cluster-state.adoc [hunk 9]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+* A device is removed or renamed from {product-title} nodes after the etcd snapshot is taken. The Local Storage Operator creates symlinks for each PV that it manages from `/dev/disk/by-id` or `/dev directories`. This situation might cause the local PVs to refer to devices that no longer exist.
```
Comment (bobfuru) [inline modules/dr-restoring-cluster-state.adoc:336]: s/`dev directories`/`dev` directories

**Review by bobfuru — COMMENTED** [review]
A few more minor edits.

File: modules/dr-restoring-cluster-state.adoc [hunk 10]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
 ----
 
 Note that it might take several minutes after completing this procedure for all services to be restored. For example, authentication by using `oc login` might not immediately work until the OAuth server pods are restarted.
+
+[id="dr-scenario-cluster-state-issues_{context}"]
+= Issues and workarounds for restoring a persistent storage state
+
+If your {product-title} cluster uses persistent storage of any form, a state of the world is typically stored outside etcd. It might be an Elasticsearch cluster running in a pod or a database running in a `StatefulSet`object. When restoring from an etcd backup, the status of the workloads in {product-title} is also restored. However, if the etcd snapshot is old, the status might be invalid or outdated.
```
Comment (mikemckiernan) [inline modules/dr-restoring-cluster-state.adoc:321]: Please add a space between `StatefulSet` and "object".

nit: Instead of "When restoring..." how do you feel about "When you restore..."?

File: modules/dr-restoring-cluster-state.adoc [hunk 11]
```diff
@@ -314,3 +314,29 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+====
+
+The following are some example scenarios that produce an out-of-date status:
+
+* MySQL database is running in a pod backed up by a PV object. Restoring {product-title} from an etcd snapshot does not bring back the volume on the storage provider, and does not produce a running MySQL pod, despite the pod repeatedly attempting to start. You must manually restore this pod by restoring the volume on the storage provider, and then editing the PV to point to the new volume.
+
+* Pod P1 is running using volume A, which is attached to node X. If the etcd snapshot is taken while another pod is using the same volume on node Y, then when the etcd restore is performed, pod P1 might not be able to start correctly due to the volume still being attached to node Y. {product-title} is not aware of the attachment, and does not automatically detach it. When this occurs, the volume must be manually detached from node Y so that the volume can attach on node X, and then pod P1 can start.
```
Comment (mikemckiernan) [inline modules/dr-restoring-cluster-state.adoc:332]: nit: "is running using" reads awkwardly. Maybe "Pod P1 uses volume A and volume A is attached to node X."

In the next sentence "is using" can be "uses". Up to you.

File: modules/dr-restoring-cluster-state.adoc (hunk 1 above)
Comment (gnufied) [reply modules/dr-restoring-cluster-state.adoc]: we missed this. @bobfuru @lpettyjo I thought we wanted to use word - "state of the cluster" ?
